dali/pipeline/operators/arithmetic/expression_tree.cc

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/pipeline/operators/arithmetic/expression_tree.h"
+#include "dali/pipeline/operators/arithmetic/arithmetic_meta.h"
+#include "dali/core/static_switch.h"
+#include "dali/pipeline/operators/arithmetic/expression_impl_cpu.h"
+
+namespace dali {
+
+namespace {
+
+// <expr> := <call>|<scalar>|<input>
+// <call> := <name> '(' <expr>* ')'
+// <name> := fun
+// <input> := &Int
+// <scalar> := $Int:Type
+
+// add(&0 mul(&1 $0:i8))
+// add(&0 rand())
+
+std::tuple<std::unique_ptr<ExpressionBase>, int> ParseExpr(const std::string &expr, int pos);
+
+int ConsumeChar(const std::string &expr, int pos, char c) {
+  if (expr[pos] == c) {
+    return pos + 1;
+  }
+  DALI_FAIL("Unrecognized token for arithmetic expression: \"" + expr[pos] + "\" at position [" +
+            std::to_string(pos) + "] in: " + expr);
+}
+
+int SkipChars(const std::string &expr, int pos, char c) {
+  while (expr[pos] == c) {
+    pos++;
+  }
+  return pos;
+}
+
+std::tuple<std::string, int> ParseName(const std::string &expr, int pos) {
+  int end_pos = pos;
+  while (std::isalnum(expr[end_pos])) {
+    end_pos++;
+  }
+  return std::make_pair(std::string(&expr[pos], &expr[end_pos]), end_pos);
+  DALI_FAIL("Unrecognized token for arithmetic expression: \"" + expr[pos] + "\" at position [" +
+            std::to_string(pos) + "] in: " + expr);
+}
+
+std::tuple<std::unique_ptr<ExpressionBase>, int> ParseCall(const std::string &expr, int pos) {
```

Comment:
I think this loooong name deserves a typedef.
```suggestion
ParseResult ParseCall(const std::string &expr, int pos) {
```

Reply:
Done